### PR TITLE
Cabal: Expose flag assignment to configure script

### PR DESCRIPTION
**Pull request by bgamari**
Here we extend the `build-type: configure` mechanism to allow the configure script to inspect the flag assignment via the `CABAL_FLAGS` environment variable, as suggested in #8564.

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: changelog.d/issue-8564 [hunk 1]
```diff
@@ -0,0 +1,5 @@
+synopsis: The `configure` script of `build-type: configure` packages now has access to the flag assignment of the package being built via the `CABAL_FLAGS` and `CABAL_FLAG_<flag>` environment variables
+packages: Caba
```
Comment (Ericson2314) [inline changelog.d/issue-8564:2]:
```suggestion
packages: Cabal
```
?

**Comment by Ericson2314** [comment]
Otherwise looks good!

**Review by gbaz — APPROVED** [review]

**Comment by Ericson2314** [comment]
How come CI is not running?

**Comment by Mikolaj** [comment]
@bgamari, @Ericson2314: do you mind if I rebase? Perhaps then CI is going to work again.

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by Mikolaj** [comment]
Validate on Windows failed due to the fragility of (our?) Windows CI. I've restarted it.

**Comment by Ericson2314** [comment]
Thanks!

**Pull request merged**